python/pyspark/ml/param/__init__.py

```diff
@@ -152,13 +152,17 @@ def isDefined(self, param):
         return self.isSet(param) or self.hasDefault(param)
 
     @since("1.4.0")
-    def hasParam(self, paramName):
+    def hasParam(self, param):
         """
-        Tests whether this instance contains a param with a given
-        (string) name.
+        Tests whether this instance contains a param.
         """
-        param = self._resolveParam(paramName)
-        return param in self.params
+        if isinstance(param, Param):
+            return hasattr(self, param.name)
```

Comment:
+1 for accepting only strings. If no strong reasons, keep consistent with Scala is the best choice.